rlpx/handshake.go

```diff
@@ -0,0 +1,138 @@
+package rlpx
+
+import (
+	"crypto/rand"
+	"errors"
+	mrand "math/rand"
+	"net"
+
+	"github.com/decred/dcrd/dcrec/secp256k1/v4"
+
+	"github.com/indexsupply/x/bint"
+	"github.com/indexsupply/x/ecies"
+	"github.com/indexsupply/x/enr"
+	"github.com/indexsupply/x/isxsecp256k1"
+	"github.com/indexsupply/x/rlp"
+)
+
+const (
+	authVersion  = 4
+	prefixLength = 2
+)
+
+type handshake struct {
+	isInitiator     bool
+	localPrvKey     *secp256k1.PrivateKey
+	localEphPrvKey  *secp256k1.PrivateKey
+	remotePubKey    *secp256k1.PublicKey
+	remoteEphPubKey *secp256k1.PublicKey
+	initNonce       []byte
+	receiverNonce   []byte
+}
+
+func newHandshake(localPrvKey *secp256k1.PrivateKey, to *enr.Record) (*handshake, error) {
+	h := &handshake{
+		remotePubKey: to.PublicKey,
+		localPrvKey:  localPrvKey,
+	}
+
+	return h, nil
+}
+
+func (h *handshake) createAuthMsg() (rlp.Item, error) {
+	var err error
+	// initialize random nonce
+	if h.initNonce == nil {
```

Comment:
Are these == nil checks for testing?

Reply:
Yep they're for testing so i can seed with a known nonce. Do you have a suggestion for a better way of dealing with it?

Reply:
I wonder if we test auth-ack (subsequently seal/unseal) in tandem instead of individually. I think that would have the same effect without having to introduce conditionals into the mainline code. WDYT?

Reply:
That makes sense. But asides from testing, I wanted to make this function idempotent so that was another reason for not randomizing the nonce if it's already present.

Reply:
Do we need this to be idempotent? I would assume that if we are making an auth message, and if that auth message doesn't work, that it would be safer to throw away the handshake struct and then try again with new state.